### DISCUSSION TITLE
SHAP in Cross Fold

### DISCUSSION THREAD

**Discussion by stevenvanwinkel** [shap/shap#1501] [Q&A]
Hi Scott, thank you for the very useful SHAP package! I am looking at using the shap values as an alternative to permutatation importance features (which I have found to be too unstable) and have a question. For the Permutation Importance, I use a cross-fold validation to derive the most important features. Is there any benefit to doing the shap values in a cross fold as well and calculate the shap values of the test set? Some implementations seem to do this, some find the shap values of the full sample without any cross-fold. What are your thoughts on this? Thank you!

**Comment by jaxondk** [comment]
I will leave my $0.02 on this.

Both permutation importance and SHAP values (in almost all cases) are derived using _interventional_ conditional expectations, which makes their explanations "true to the model", but not "true to the data." This is because both methods intervene by assigning values to simulate "missing" features, which can generate instances off of the data manifold. But the values will explain what features the model actually is using to predict, whether or not those features are actually as important in the actual patterns in the data. (See [this github discussion](https://github.com/slundberg/shap/discussions/1538#discussioncomment-109573) and the [True to the model or true to the data](https://arxiv.org/pdf/2006.16234.pdf) paper.)

This has an important implication for cross-fold validation. If you have 10 folds, you will have 10 different models. These models may vary somewhat, depending on the type of model you're using (for example, DL models have a high variance even when re-training on the same data) and the consistency of your data across all the folds. And since SHAP (and PI, for that matter) are both "true to the model", you may get fairly different explanations for one fold vs. another, since each fold has its own model that may behave differently.

This is just something to be aware of if you're combining shap values from multiple models, like you would be in cross-fold validation. It may be worth examining each fold's explanations before combining them to see if they vary significantly

I look forward to hearing from others as well! Or perhaps you've made some progress on this yourself @stevenvanwinkel since posting about it?
